utils/cmake/dreamcast.toolchain.cmake

```diff
@@ -0,0 +1,106 @@
+# CMake Toolchain file for targeting the Dreamcast or NAOMI with CMake.
+#  (c)2023 Falco Girgis, Luke Benstead
+#
+# This file is to be passed to CMake when compiling a regular CMake project
+# to cross-compile for the Dreamcast, using the KOS environment and settings.
+#
+#     cmake /path/to/src -DCMAKE_TOOLCHAIN_FILE=${KOS_BASE}/kos/utils/cmake/dreamcast.toolchain.cmake
+#
+#   or alternatively:
+#
+#     cmake /path/to/src -DCMAKE_TOOLCHAIN_FILE=${KOS_CMAKE_TOOLCHAIN}
+#
+# The toolchain enables the options required for the compiler to use the 
+# special SH4 instructions for sin/cos and invSqrt. Note that even when 
+# these options are present, they are not activated unless -ffast-math 
+# is also provied. Typically you will want to provide a project-level 
+# option for enabling this flag, which will make the toolchain's SH4 
+# instruction flags also take effect.
+#
+# Frame pointers are enabled in debug builds as these are required for 
+# stack traces and GDB. They are disabled in release.
+# 
+# The original toolchain file was created by Kazade for the Simulant 
+# engine who has graciously allowed the rest of the scene to warez it. 
+#
+# Special thanks to Matt Godbolt of Compiler Explorer, for supporting
+# the SH toolchains, which allowed us to figure out the precise magical
+# incantations of compiler flags for fast math.  
+
+##### Configure CMake System #####
+set(CMAKE_SYSTEM_NAME Generic)
+set(CMAKE_SYSTEM_VERSION 1)
+set(PLATFORM_DREAMCAST TRUE)
+
+##### Configure Cross-Compiler #####
+set(CMAKE_CROSSCOMPILING TRUE)
+set(CMAKE_C_COMPILER $ENV{KOS_CC_BASE}/bin/sh-elf-gcc)
+set(CMAKE_CXX_COMPILER $ENV{KOS_CC_BASE}/bin/sh-elf-g++)
+
+set(CMAKE_EXECUTABLE_SUFFIX ".elf")
+
+set(CMAKE_FIND_LIBRARY_SUFFIXES ".a")
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+
+##### Add Platform-Specific #defines #####
+ADD_DEFINITIONS(
+    -D__DREAMCAST__
+    -D_arch_dreamcast
+)
+
+if($ENV{KOS_SUBARCH} MATCHES naomi)
+    ADD_DEFINITONS(
+        -D__NAOMI__
+        -D_arch_sub_naomi
+    )
+else()
+    ADD_DEFINITIONS(-D_arch_sub_pristine)
+endif()
+
+##### Configure Build Flags #####
+add_compile_options(-ml -m4-single-only -ffunction-sections -fdata-sections)
+
+set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -DFRAME_POINTERS -mfsrra -mfsca -fno-omit-frame-pointer")
```

Comment:
Can fix the repetition of these flags using generator expressions.

```
##### Configure Build Flags #####
add_compile_options(-ml -m4-single-only -ffunction-sections -fdata-sections)

set(ENABLE_DEBUG_FLAGS   $<OR:$<CONFIG:Debug>,$<CONFIG:RelWithDebInfo>>)
set(ENABLE_RELEASE_FLAGS $<OR:$<CONFIG:Release>,$<CONFIG:MinSizeRel>>)

add_compile_options(
    "$<${ENABLE_DEBUG_FLAGS}:-DFRAME_POINTERS;-fno-omit-frame-pointer>"
    "$<${ENABLE_RELEASE_FLAGS}:-fomit-frame-pointer>"
    )

set(CMAKE_ASM_FLAGS "")
set(CMAKE_ASM_FLAGS_RELEASE "")
```

For handling the SH4 Math flags the best I could come up with was having an include-able helper file with a function to add the math flags to the compile. The Vita SDK includes a similar helper file https://github.com/vitasdk/samples/blob/master/hello_world/CMakeLists.txt#L21. 

The issue being that the `-mfsrra` and `-mfsca` flags are not available on the GCC4.7.4 (legacy) toolchain. We can't check if they're avaiable in the toolchain file, as CMake doesn't run any checks on the compiler until after the toolchain file finishes loading, so we don't know the GCC version or supported compiler flags inside the toolchain file. 

This helper file could be included alongside the toolchain file in KOS. See https://github.com/cepawiel/tutorials/blob/main/dreamcast.cmake#L117 for an example. I do believe I'm missing a few of the additional math flags required, just was a quick example based on the flags shown here. Also allows us to include other helper functions like creating a romdisk and a CDI (also in linked file).